pkg/solidity-utils/contracts/helpers/BaseSplitCodeFactory.sol

```diff
@@ -0,0 +1,208 @@
+
+    address private immutable _creationCodeStorageB;
+    uint256 private immutable _creationCodeSizeB;
+
+    /**
+     * @dev The creation code of a contract Foo can be obtained inside Solidity with `type(Foo).creationCode`.
+     */
+    constructor(bytes memory creationCode) {
+        uint256 creationCodeSize = creationCode.length;
+
+        // We are going to deploy two contracts: one with approximately the first half of `creationCode`'s contents
+        // (A), and another with the remaining half (B).
+        // We store the lengths in both immutable and stack variables, since immutable variables cannot be read during
+        // construction.
+        uint256 creationCodeSizeA = creationCodeSize / 2;
+        _creationCodeSizeA = creationCodeSizeA;
+
+        uint256 creationCodeSizeB = creationCodeSize - creationCodeSizeA;
+        _creationCodeSizeB = creationCodeSizeB;
+
+        // To deploy the contracts, we're going to use `CodeDeployer.deploy()`, which expects a memory array with
+        // the code to deploy. Note that we cannot simply create arrays for A and B's code by copying or moving
+        // `creationCode`'s contents as they are expected to be very large (> 24kB), so we must operate in-place.
+
+        // Memory: [ code length ] [ A.data ] [ B.data ]
+
+        // Creating A's array is simple: we simply replace `creationCode`'s length with A's length. We'll later restore
+        // the original length.
+
+        bytes memory creationCodeA;
+        assembly {
+            creationCodeA := creationCode
+            mstore(creationCodeA, creationCodeSizeA)
+        }
+
+        // Memory: [ A.length ] [ A.data ] [ B.data ]
+        //         ^ creationCodeA
+
+        _creationCodeStorageA = CodeDeployer.deploy(creationCodeA);
+
+        // Creating B's array is a bit more involved: since we cannot move B's contents, we are going to create a 'new'
+        // memory array starting at A's last 32 bytes, which will be replaced with B's length. We'll back-up this last
+        // byte to later restore it.
+
+        bytes memory creationCodeB;
+        bytes32 lastByteA;
+
+        assembly {
+            // `creationCode` points to the array's length, not data, so by adding A's length to it we arrive at A's
+            // last 32 bytes.
+            creationCodeB := add(creationCode, creationCodeSizeA)
+            lastByteA := mload(creationCodeB)
+            mstore(creationCodeB, creationCodeSizeB)
+        }
+
+        // Memory: [ A.length ] [ A.data[ : -1] ] [ B.length ][ B.data ]
+        //         ^ creationCodeA                ^ creationCodeB
+
+        _creationCodeStorageB = CodeDeployer.deploy(creationCodeB);
+
+        // We now restore the original contents of `creationCode` by writing back the original length and A's last byte.
+        assembly {
+            mstore(creationCodeA, creationCodeSize)
+            mstore(creationCodeB, lastByteA)
+        }
```

Comment:
Why do you need to restore these contents?

Reply:
I think this can be simplified in a few lines if you don't care that much about the final memory layout

Reply:
We don't _need_ to restore them - in fact, in the expected use case where we do `SplitCodeFactory(type(Foo).creationCode)` the reference to the original creation code will be lost anyway.

But restoring the original value is so simple that I'd rather we don't mutate it at all and can not worry about it.